### ISSUE TITLE
Wazuh App - Authorization Exception

### ISSUE THREAD

**Issue by mathieu83470** [wazuh/wazuh-kibana-app#1203]
Hello,

I have exactly the same issue of this case (https://github.com/wazuh/wazuh-kibana-app/issues/968) but i can't disable Xpack , anyone know to resolve it?
In kibana.yml => I use "elastic" user
In kibana UI => Logged with same "elastic" user

Error message : {"message":"2001 - [security_exception] action [indices:data/read/search] is unauthorized for user [kibana]","code":2001,"statusCode":500}

In kibana log - Initialization:
![image](https://user-images.githubusercontent.com/46820664/52118720-61e4d200-2617-11e9-9649-9898d3ab1e35.png)

In kibana log - Accessing to wazuh-app:
![image](https://user-images.githubusercontent.com/46820664/52119071-40381a80-2618-11e9-82c1-495415930d63.png)

Kibana version : 6.4.0
Wazuh-app version : Tested with 3.6.0 and  3.6.1 with same results

Thanks for your help

**Comment by mathieu83470** [comment]
Hi @jesusgn90 

Thanks but for now I can't upgrade Elastick stack to 6.6!
I'm exactly in the same configuration of @fl4nker (https://github.com/wazuh/wazuh-kibana-app/issues/968) except i can't disable xpack for security reasons.



This is my config :1st_place_medal: 
![image](https://user-images.githubusercontent.com/46820664/52122952-fbff4700-2624-11e9-8dee-068d0eed1124.png)

![image](https://user-images.githubusercontent.com/46820664/52123183-952e5d80-2625-11e9-9d84-c9a812725551.png)

**Comment by jesusgn90** [comment]
Uncomment the next line:

https://github.com/wazuh/wazuh-kibana-app/blob/bb9b1de686b75d73bf36226ac63e858f69f05f4c/config.yml#L75

The file is located under `/usr/share/kibana/plugins/wazuh/config.yml`

So it should look like this:
```yml
xpack.rbac.enabled: true 
```

Restart Kibana:

```
systemctl restart kibana
```

Regards

**Comment by mathieu83470** [comment]
Unfortunately, it was already done with same results....

![image](https://user-images.githubusercontent.com/46820664/52123870-aed0a480-2627-11e9-9358-ad119d29436a.png)

**Comment by jesusgn90** [comment]
Hello again @mathieu83470 ,

Please, execute the next `curl` commands:

```
curl -X GET "localhost:9200/_xpack/license"
curl -X GET "localhost:9200/_xpack/license/trial_status"
```

Where `localhost` is your Elasticsearch node that Kibana is pointing to.

Regards

**Comment by mathieu83470** [comment]
Hello @jesusgn90 

Thanks for your time! We have a gold license...

![image](https://user-images.githubusercontent.com/46820664/52198185-44f00f00-2862-11e9-97a7-83d50129713f.png)


![image](https://user-images.githubusercontent.com/46820664/52198154-37d32000-2862-11e9-980c-66fed327c881.png)

Regards

**Comment by jesusgn90** [comment]
Hi @mathieu83470 I think I got it, you are using a Wazuh app a version older than a fixed app we released for 6.4.2 (https://github.com/wazuh/wazuh-kibana-app/pull/998). You are probably facing an issue with X-Pack we have for 3.6 and 6.4.0 / 6.4.1

Edit the file under `/usr/share/kibana/plugins/wazuh/server/lib/elastic-wrapper.js`

Look for a function named `updateWazuhIndexDocument`

Replace the next line:

```js
const data = await this.elasticRequest.callWithRequest(req, 'search', {
```

with this:

```js
const data = await this.elasticRequest.callWithInternalUser('search', {
```

Save it and restart Kibana:

```
# chown -R kibana:kibana /usr/share/kibana/plugins
# systemctl restart kibana
```

Best regards

**Comment by mathieu83470** [comment]
@jesusgn90 

I changed 2 lines with no changes after a restart of kibana !

![image](https://user-images.githubusercontent.com/46820664/52201483-cdbf7880-286b-11e9-8526-fdd106237fc1.png)



![image](https://user-images.githubusercontent.com/46820664/52201534-fd6e8080-286b-11e9-8f80-685f72eb1fd7.png)

**Comment by jesusgn90** [comment]
Those lines are not the desired lines, the line I was talking about is https://github.com/wazuh/wazuh-kibana-app/blob/bb9b1de686b75d73bf36226ac63e858f69f05f4c/server/lib/elastic-wrapper.js#L498, that line must be `const data = await this.elasticRequest.callWithInternalUser('search', {` for your version. So please, undo your changes and follow my instructions from https://github.com/wazuh/wazuh-kibana-app/issues/1203#issuecomment-460175633

Kind regards,
Jesús

**Comment by mathieu83470** [comment]
Ok, sorry for the misunderstanding.
I did it with no effects... :-(

![image](https://user-images.githubusercontent.com/46820664/52202356-90101f00-286e-11e9-97c2-08ee14a4d6ef.png)


![image](https://user-images.githubusercontent.com/46820664/52202213-2728a700-286e-11e9-98ee-9eee02b9f4df.png)

**Comment by jesusgn90** [comment]
Hello again @mathieu83470 , sorry for the late response.

Please, can you paste the output from the next commands?

```
cat /usr/share/kibana/plugins/wazuh/server/lib/elastic-wrapper.js | grep updateWazuhIndex -A16
cat /etc/kibana/kibana.yml # Replace passwords with ****
```

Thanks!

**Comment by mathieu83470** [comment]
Hello @jesusgn90 

Thanks for your help

![image](https://user-images.githubusercontent.com/46820664/52344871-2df51c80-2a1c-11e9-8d70-e05c0e0a8732.png)

![image](https://user-images.githubusercontent.com/46820664/52345163-c7243300-2a1c-11e9-8b48-1ab139522abe.png)

**Comment by jesusgn90** [comment]
All seems fine, the issue comes from the internal app indices (`.wazuh` and `.wazuh-version`).

Please paste the output of the next command:

```
curl localhost:9200/.wazuh,.wazuh-version/_settings?pretty
```

Example output:

```js
{
  ".wazuh" : {
    "settings" : {
      "index" : {
        "number_of_shards" : "1",
        "provided_name" : ".wazuh",
        "creation_date" : "1549060666239",
        "number_of_replicas" : "0",
        "uuid" : "nA61yOZyQIKLkxYH9JxirQ",
        "version" : {
          "created" : "6060099"
        }
      }
    }
  },
  ".wazuh-version" : {
    "settings" : {
      "index" : {
        "creation_date" : "1549060666430",
        "number_of_shards" : "1",
        "number_of_replicas" : "0",
        "uuid" : "DRRHumAYRsayUEv5gSWF2A",
        "version" : {
          "created" : "6060099"
        },
        "provided_name" : ".wazuh-version"
      }
    }
  }
}
```

I'm asking for that because Elasticsearch sometimes blocks certain indices, in that case you'll see an entry like the next one:

```js
        "blocks" : {
          "read_only_allow_delete" : "true"
        },
```

If it appears in your output, use this comment to unlock those indices:

```js
curl -XPUT 'http://localhost:9200/_settings' -H 'Content-Type: application/json' -d' { "index": { "blocks": { "read_only_allow_delete": "false" } } } '
```

Where `localhost` is always the Elasticsearch host you provided in `kibana.yml`.

Regards

**Comment by mathieu83470** [comment]
They are not created...

![image](https://user-images.githubusercontent.com/46820664/52351991-0efe8680-2a2c-11e9-8f63-8bbfd5af3642.png)

At startup I have "Error creating index" error logs

![](https://user-images.githubusercontent.com/46820664/52202356-90101f00-286e-11e9-97c2-08ee14a4d6ef.png)

**Comment by jesusgn90** [comment]
That's right, I was thinking about the `.wazuh` index... well, at this point, I need to research about your issue more in deep, let me debug a bit your version in a lab environment, adjusting even more my settings in order to have a very similar environment because something weird is happening under the hood that I'm not getting so clear. 

I'll get back to you as soon as possible.

Regards

**Comment by mathieu83470** [comment]
Thanks a lot

Maybe we can also try to upgrade ELK stack to 6.6 and re-install new version of wazuh app.

For your information, to install wazuh app I had to follow this procedure : https://github.com/wazuh/wazuh-kibana-app/issues/664#issuecomment-430690314
(after hours blocked on "Optimizing and caching bundles..." screen)

**Comment by jesusgn90** [comment]
Well, in any case, that's not the solution, I'm building a similar environment so I can debug properly what's happening to you. 

We are not planning to backport any change to 6.4.X but that version worked for other users (just replacing the line a suggested some comments ago), so I want to hunt what's happening to your environment.

**Comment by jesusgn90** [comment]
In the meantime, could you try to open a new "incognito" window on your browser? or just "clear all site data" in order to avoid a cookie/session issue related to Kibana itself.

**Comment by mathieu83470** [comment]
Same problem in private navigation 

![image](https://user-images.githubusercontent.com/46820664/52354895-926ea680-2a31-11e9-9f0e-6b32679a682a.png)

**Comment by jesusgn90** [comment]
Follow up @mathieu83470 :

![image](https://user-images.githubusercontent.com/8860227/52355792-50466480-2a33-11e9-8404-e3f147459045.png)


Working app using simple auth and the user `elastic` both sides (`kibana.yml` + UI login).

Wazuh 3.6.1 - Elastic 6.4.0

Now I'm going to apply extra settings from your configuration.

**Comment by jesusgn90** [comment]
Hello again @mathieu83470 , I've been testing your configuration, this is my own:

```yml
server.host: "0.0.0.0"
xpack.security.enabled: true
xpack.security.encryptionKey: "something_at_least_32_characters"
xpack.security.secureCookies: true
elasticsearch.url: "http://172.16.1.2:9200"
elasticsearch.username: "elastic"
elasticsearch.password: "elastic"
server.ssl.enabled: true
server.ssl.certificate: /tmp/host.crt
server.ssl.key: /tmp/host.key
logging.dest: /var/log/kibana/kibana.log
logging.verbose: true
elasticsearch.ssl.certificateAuthorities: [ "/tmp/hostCA.crt" ]
```

All is working for me, can not reproduce your situation.

Tried to access the Wazuh app, deleted multiple times `.wazuh` and `.wazuh-version`... So my last thought is about a reverse proxy (NGINX or something similar) and headers being missing when passing through the proxy... The way how Kibana server is communicating with Elastic server.

**Comment by mathieu83470** [comment]
Hello @jesusgn90 

Thanks again for your help.
I think we will try to upgrade ELK stack quickly, I cross fingers it will work with last version.  
I let you know

**Comment by jesusgn90** [comment]
Ok @mathieu83470 , closing this ticket since you are not planning to continue the troubleshooting. Open a new ticket if you found any more troubles.

Regards!
